### DISCUSSION TITLE
AOP에 대한 설명

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#197] [Spring Core]
.

**Comment by jimi567** [comment]
### AOP(Aspect-Oriented Programming)

+ AOP란?
    - 객체 지향 프로그래밍 패러다임을 보완하는 기술
    - **메소드나 객체의 기능을 핵심 관심사와 공통 관심사로 나누어 프로그래밍 하는 것**
    - 핵심 관심사? 각 객체가 가져야 할 본래의 기능
    - 공통 관심사? 여러 객체에서 공통적으로 사용되는 코드 
--> 여러 개의 클래스에서 반복해서 사용하는 코드가 있다면 해당 코드를 모듈화 하여 공통 관심사로 분리

### AOP 용어 

+ Aspect 

공통적인 기능을 모듈화 한 것을 의미

+ Target 

Aspect가 적용될 대상을 의미, 메소드 또는 클래스 레벨에 적용

+ Join Point 

Aspect 이 적용될 수 있는 시점을 의미, 메소드 실행 전, 후 등

+ Advice

Aspect 기능을 정의한 것으로 메소드 실행 전, 후 예외 처리 발생 시 실행되는 코드를 의미 

+ Point cut

Advice를 적용할 메소드의 범위를 지정하는 것을 의미 특정 패키지, 클래스, 메소드 이름 또는 정규 표현식을 사용하여 범위를 지정할 수 있음 

+ Weaving

컴파일 시점, 클래스 로딩 시점, 런타임 시점 중 어느 때에 Advice적용할지 결정하는 프로세스이다. 스프링 AOP는 런타임 시점에 프록시를 생성하여 Advice 적용

###  AOP 관련 어노테이션 

+ @Aspect  

해당 클래스를 Aspect 로 사용하겠다는 것을 명시

+ @Before

대상 메서드가 실행되기 전에 Advice을 실행

+ @After

메서드가 실행된 후에 Advice를 실행

+ @AfterReturning

대상 메서드가 정상적으로 실행되고 반환된 후에 Advice를 실행

+ @AfterThrowing

메서드에서 예외가 발생했을 때 Advice를 실행

+ @Around

메서드 실행 전, 후 또는 예외 발생 시에 Advice를 실행

### AOP를 쓰는 이유

  - 모든 메소드마다 공통적으로 실행되는 로직이 있을때
  - 예를 들어, 각 메소드 마다 소요되는 시간을 측정하고 싶을 때,
  - **공통관심사항, 핵심 관심 사항을 분리하기 위해**



### 아래는 영한님 강의에서 AOP를 적용하는 예시 

+ AOP를 사용하지않고, 공통적으로 실행되는 로직을 추가할때 문제
예를 들어, MemberService에 각 메소드마다 시간을 측정하는 로직을 추가함
    - 회원가입, 회원 조회에 시간을 측정하는 기능은 핵심 관심 사항이 아님
    - 시간을 측정하는 로직은 공통 관심 사항
    - 시간을 측정하는 로직과 핵심 비즈니스의 로직이 섞여서 유지보수가 어려움
    - 시간을 측정하는 로직을 별도의 공통 로직으로 만들기 매우 어려움
    - 시간을 측정하는 로직을 변경할 때 모든 로직을 찾아가면서 변경해야됨(시간 측정 단위를 s에서 m/s로 변경할때 등)

+ AOP 적용
    -  AOP: Aspect Oriented Programming
    -  공통 관심 사항(cross-cutting concern) vs 핵심 관심 사항(core concern) 분리
    
TimeTraceAop.java
```java
@Component
@Aspect
public class TimeTraceAop {
    @Around("execution(* hello.hellospring..*(..))")
    public Object execute(ProceedingJoinPoint joinPoint) throws Throwable {
        long start = System.currentTimeMillis();
        System.out.println("START: " + joinPoint.toString());
        try {
            return joinPoint.proceed();
        } finally {
            long finish = System.currentTimeMillis();
            long timeMs = finish - start;
            System.out.println("END: " + joinPoint.toString()+ " " + timeMs +
            "ms");
           }
        }
    }
```
+ 문제점 해결
    - 회원가입, 회원 조회등 핵심 관심사항과 시간을 측정하는 공통 관심 사항을 분리
    - 시간을 측정하는 로직을 별도의 공통 로직으로 만듦
    - 핵심 관심 사항을 깔끔하게 유지가능
    - 공통 관심 사항에 변경이 생기면 Aop로직만 변경하면 됨
    - 원하는 적용 대상 선택 가능(@Around의 경로 변경)
---
### AOP 동작 설명
기존 

![image](https://user-images.githubusercontent.com/28584160/215317166-a0514b01-9bbd-49a8-82cb-c320c6967029.png)

AOP 적용

![image](https://user-images.githubusercontent.com/28584160/215317193-efd18154-e787-490f-842d-938611dc3304.png)

프록시라는 가짜 스프링 빈을 등록하여 실제 스프링 빈에 DI하기전에 프록시 빈을 거쳐가도록 동작하게됨 

---
번외

스프링 AOP는 JDK 동적 프록시와 CGLIB 프록시를 사용해 AOP를 구현함   
이 부분은 정확히 맞는 내용인지 몰라서 [참고 자료](https://gmoon92.github.io/spring/aop/2019/04/20/jdk-dynamic-proxy-and-cglib.html)

**Comment by GIVEN53** [comment]
![image](https://github.com/2024-woowacourse-study/level-interview/assets/101033262/80bc9c7a-c271-4555-9be8-796a2b28321f)

비즈니스 로직과 같이 애플리케이션의 주목적을 달성하기 위한 관심사를 **핵심 관심 사항(Core concern)** 이라고 하고 로깅, 보안, 트랜잭션과 같이 공통적인 기능에 대한 관심사를 **공통 관심 사항(Cross-cutting concern)** 이라고 합니다. AOP는 핵심 관심 사항에서 공통 관심 사항 분리를 목표로 하는 소프트웨어 페러다임입니다.

> 공통 관심 사항은 따로 모듈화해서 재사용할게~

### AOP 관련 용어
#### 1. Aspect
![image](https://github.com/2024-woowacourse-study/level-interview/assets/101033262/45d23d1e-2e6b-40b8-9fd8-359b56ec7284)

여러 객체에 공통으로 적용되는 부가 기능을 모듈화한 것을 말하며 Advice와 PointCut을 합친 개념입니다.

#### 2. Advice (what)
Transaction, Security, Logging와 같이 각각의 부가 기능을 구현한 코드를 가리킵니다.

#### 3. Target (where)
Aspect를 적용할 클래스, 메서드 등을 의미하며 PointCut으로 결정됩니다.

#### 4. JointPoint (when)
클래스 초기화, 객체 인스턴화, 메서드 실행 전∙후, 필드 접근, 예외 발생과 같은 애플리케이션 실행 흐름에서의 특정 포인트를 의미합니다.

#### 5. PointCut
JoinPoint 중에서 Advice를 어디에 적용할지 필터링합니다.

#### 6. Weaving
PointCut으로 결정한 Target의 JoinPoint에 Advice를 엮은 것으로, Target에 Aspect를 적용해서 새로운 AOP 객체를 생성하는 과정입니다.

#### 7. AOP Proxy
AOP 기능을 구현하기 위해 생성된 프록시 객체이며 Spring 프레임워크에서 AOP Proxy는 JDK Dynamic Proxy 또는 CGLIB Proxy입니다.

### AOP 적용 방식
![image](https://github.com/2024-woowacourse-study/level-interview/assets/101033262/b2aaa4e6-81c9-44b3-91e0-439539f3c6f3)

AOP 적용 방식은 컴파일 타임, 클래스 로드 타임, 런타임 3가지가 있습니다. 하지만 컴파일 타임, 클래스 로드 타임 방식은 AspectJ 프레임워크를 직접 사용해야 하는 번거로움이 있어 주로 **런타임 방식**인 Spring AOP를 사용합니다.
Spring AOP는 프록시를 사용하기 때문에 메서드 실행 시점에만 AOP 적용, Bean에만 적용, 내부 메서드 호출 시 적용되지 않는 등 여러가지 제약이 있습니다.

### 추가 키워드
- AOP 적용 방식 3가지 차이
- Spring AOP vs AspectJ
